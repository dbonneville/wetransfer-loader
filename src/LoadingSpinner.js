import React, { useState, useEffect, useRef } from "react";
import "./App.css";
import "./LoadingInfo";
import LoadingInfo from "./LoadingInfo";

function useInterval(callback, delay) {
  const savedCallback = useRef();

  // remember the latest callback
  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // set up the interval
  useEffect(() => {
    let id = setInterval(() => {
      savedCallback.current();
    }, delay);
    return () => clearInterval(id);
  }, [delay]);
}

export default function LoadingSpinner() {
  const [counter, setCounter] = useState(0);
  const [isPlaying, setIsPlaying] = useState(false);

  /* toggle loading animation */
  const handlePlayToggle = () => {
    setCounter(0);
    setIsPlaying((prevPlaying) => !prevPlaying);
  };

  useInterval(() => {
    if (counter < 100 && isPlaying) {
      setCounter(counter + 1);
    } else if (counter === 100) {
      setIsPlaying(false);
    }
  }, 80);

  const leftRotation = () => {
    if (counter <= 50) {
      let leftDeg = counter * 3.6;
      return leftDeg;
    }
  };

  const rightRotation = () => {
    if (counter > 50) {
      let rightDeg = counter * 3.6 - 180;
      return rightDeg;
    }
  };

  const fullRotation = () => {
    let fullDeg = counter * 3.6 - 90;
    return fullDeg;
  };

  const counterPosition = () => {
    if (counter < 10) {
      return 173;
    } else if (counter >= 10 && counter < 100) {
      return 155;
    } else {
      return 140;
    }
  };

  const pctPosition = () => {
    if (counter < 10) {
      return 155;
    } else if (counter >= 10 && counter < 100) {
      return 173;
    } else {
      return 185;
    }
  };

  console.log("isPlaying: " + isPlaying);

  return (
    <>
      <div className="container">
        <div className="number" style={{ left: `${counterPosition()}px` }}>
          {counter}
        </div>
        <div className="number-pct" style={{ left: `${pctPosition()}px` }}>
          %
        </div>
        <div
          className="circular"
          style={{
            animation: isPlaying ? `spinAll 2s linear both infinite` : ``,
          }}
        >
          <div className="spin-container">
            <div className="inner"></div>
            <div className="circle">
              <div
                className="dot"
                style={{
                  transform: `rotate(${fullRotation()}deg)`,
                  display: isPlaying ? `block` : `none`,
                }}
              >
                <span></span>
              </div>
              {isPlaying && (
                <div className="staticDot">
                  <span></span>
                </div>
              )}
              <div className="bar left">
                <div
                  className="progress"
                  style={{
                    transform: `rotate(${leftRotation()}deg)`,
                    display: isPlaying ? `block` : `none`,
                  }}
                ></div>
              </div>
              <div id="barRight" className="bar right">
                <div
                  className="progress"
                  style={{
                    transform: isPlaying
                      ? `rotate(${rightRotation()}deg)`
                      : `rotate(0deg)`,
                    display: isPlaying ? `block` : `none`,
                  }}
                ></div>
              </div>
            </div>
          </div>
        </div>
        <LoadingInfo />
        <div className="btnContainer">
          <button className="toggleBtn" onClick={handlePlayToggle}>
            {isPlaying ? `End` : `Start`}
          </button>
        </div>
      </div>
    </>
  );
}
