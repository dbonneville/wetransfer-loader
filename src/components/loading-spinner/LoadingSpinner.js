import React from "react";
import "../../App.css";

export default function LoadingSpinner({
  counter: percentage,
  isPlaying,
  spinning,
}) {
  //
  const leftRotation = () => {
    // let leftDeg = 0;
    if (percentage <= 50) {
      let leftDeg = percentage * 3.6;
      return leftDeg;
    }
  };

  const rightRotation = () => {
    if (percentage > 50) {
      let rightDeg = percentage * 3.6 - 180;
      return rightDeg;
    }
  };

  const fullRotation = () => {
    let fullDeg = percentage * 3.6 - 90;
    return fullDeg;
  };

  const counterPosition = () => {
    if (percentage < 10) {
      return 173;
    } else if (percentage >= 10 && percentage < 100) {
      return 155;
    } else {
      return 140;
    }
  };

  const pctPosition = () => {
    if (percentage < 10) {
      return 155;
    } else if (percentage >= 10 && percentage < 100) {
      return 173;
    } else {
      return 185;
    }
  };

  return (
    <>
      <div
        className="number"
        data-testid="number"
        style={{ left: `${counterPosition()}px` }}
      >
        {percentage}
      </div>
      <div className="number-pct" style={{ left: `${pctPosition()}px` }}>
        %
      </div>
      <div
        className="circular"
        style={{
          animation: spinning ? `spinAll 2s linear both infinite` : ``,
        }}
      >
        <div className="spin-container">
          <div className="inner"></div>
          <div className="circle">
            <div
              className="dot"
              style={{
                transform: `rotate(${fullRotation()}deg)`,
                display: isPlaying ? `block` : `none`,
              }}
            >
              <span></span>
            </div>
            {isPlaying && (
              <div className="staticDot">
                <span></span>
              </div>
            )}
            <div className="bar left">
              <div
                className="progress"
                style={{
                  transform: `rotate(${leftRotation()}deg)`,
                  display: isPlaying ? `block` : `none`,
                }}
              ></div>
            </div>
            <div id="barRight" className="bar right">
              <div
                className="progress"
                style={{
                  transform: isPlaying
                    ? `rotate(${rightRotation()}deg)`
                    : `rotate(0deg)`,
                  display: isPlaying ? `block` : `none`,
                }}
              ></div>
            </div>
          </div>
        </div>
      </div>
    </>
  );
}
