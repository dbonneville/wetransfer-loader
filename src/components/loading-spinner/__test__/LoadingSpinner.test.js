import React from "react";
import renderer from "react-test-renderer";

import LoadingSpinner from "../LoadingSpinner";

it("renders 55% in the spinner when not playing", () => {
  const component = renderer.create(<LoadingSpinner counter={55} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders 55% in the spinner", () => {
  const component = renderer.create(<LoadingSpinner counter={55} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
