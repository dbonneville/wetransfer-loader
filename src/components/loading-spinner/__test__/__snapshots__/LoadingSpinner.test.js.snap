// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 55% in the spinner 1`] = `
Array [
  <div
    className="number"
    data-testid="number"
    style={
      Object {
        "left": "155px",
      }
    }
  >
    55
  </div>,
  <div
    className="number-pct"
    style={
      Object {
        "left": "173px",
      }
    }
  >
    %
  </div>,
  <div
    className="circular"
    style={
      Object {
        "animation": "",
      }
    }
  >
    <div
      className="spin-container"
    >
      <div
        className="inner"
      />
      <div
        className="circle"
      >
        <div
          className="dot"
          style={
            Object {
              "display": "none",
              "transform": "rotate(108deg)",
            }
          }
        >
          <span />
        </div>
        <div
          className="bar left"
        >
          <div
            className="progress"
            style={
              Object {
                "display": "none",
                "transform": "rotate(0deg)",
              }
            }
          />
        </div>
        <div
          className="bar right"
          id="barRight"
        >
          <div
            className="progress"
            style={
              Object {
                "display": "none",
                "transform": "rotate(0deg)",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`renders 55% in the spinner when not playing 1`] = `
Array [
  <div
    className="number"
    data-testid="number"
    style={
      Object {
        "left": "155px",
      }
    }
  >
    55
  </div>,
  <div
    className="number-pct"
    style={
      Object {
        "left": "173px",
      }
    }
  >
    %
  </div>,
  <div
    className="circular"
    style={
      Object {
        "animation": "",
      }
    }
  >
    <div
      className="spin-container"
    >
      <div
        className="inner"
      />
      <div
        className="circle"
      >
        <div
          className="dot"
          style={
            Object {
              "display": "none",
              "transform": "rotate(108deg)",
            }
          }
        >
          <span />
        </div>
        <div
          className="bar left"
        >
          <div
            className="progress"
            style={
              Object {
                "display": "none",
                "transform": "rotate(undefineddeg)",
              }
            }
          />
        </div>
        <div
          className="bar right"
          id="barRight"
        >
          <div
            className="progress"
            style={
              Object {
                "display": "none",
                "transform": "rotate(0deg)",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>,
]
`;
